### PR TITLE
Native csi camera integration

### PR DESCRIPTION
**Pull request by nayanavenkataramana**

### CODE REVIEW

File: src/CsiCameraDevice.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+      Terminate.
+    */
+    void CsiCameraDevice::terminate(void)
+    {
+        LINF_(TAG, "CameraDevice terminate");
+        if(m_pGPIOClass)
+        {
+            GPIOControl_release(m_pGPIOClass);
+        }
+        /* Venkat disconnectWlConnection(); */
```
Comment (NitinGote) [inline src/CsiCameraDevice.cpp]: Invalid comment.

File: src/DeviceController.cpp [hunk 2]
```diff
@@ -92,6 +93,19 @@ namespace earlyapp
                 GstVideoDevice::getInstance() : pVid;
             m_pCam = (pVid == nullptr) ?
```
Comment (litcoder) [inline src/DeviceController.cpp]: Couldn't this be removed due to the else statement below?

Comment (litcoder) [inline src/DeviceController.cpp]: This line is redundant due to the else statement below.

File: src/DeviceController.cpp [hunk 3]
```diff
@@ -92,6 +93,19 @@ namespace earlyapp
...
                 GstCameraDevice::getInstance() : pCam;
+	     
+	    if(m_pConf->useCsicam())
```
Comment (litcoder) [inline src/DeviceController.cpp]: With new devices coming in, here will require more and more ugly if-else statements which makes this difficult to maintain. Why don't you consider a factory pattern that support multiple camera configurations?

File: src/DeviceController.cpp [hunk 4]
```diff
@@ -111,8 +125,12 @@ namespace earlyapp
         dmesgLogPrint("EA: Waiting for Wayland socket...");
 #endif
         // Wait for the Wayland.
-        if(bWaitWL)
+	if(!m_pConf->useCsicam())
```
Comment (litcoder) [inline src/DeviceController.cpp]: If you have a good reason to postpone a waitForWayland() method call after thread creations, why don't you keep one single code? "Wait for waylaid if the camera configuration is CSI" doesn't make sense to me..

File: src/DeviceController.cpp [hunk 5]
```diff
@@ -131,6 +149,9 @@ namespace earlyapp
         dmesgLogPrint("EA: Devices initialized");
 #endif
         m_bInit = true;
+
+	if(m_pConf->useCsicam())
```
Comment (litcoder) [inline src/DeviceController.cpp]: See above comment.

File: src/DeviceController.cpp (hunk 2 above)
Comment (nayanavenkataramana) [reply src/DeviceController.cpp]: yes, will remove this line because using in else statement

File: src/DeviceController.cpp (hunk 4 above)
Comment (nayanavenkataramana) [reply src/DeviceController.cpp]: This Csi camera path integrated in clear Linux, here we are using "useCsicam" flags  and making sure that changes doesn't effect the SOS behaviour also.

File: src/DeviceController.cpp (hunk 5 above)
Comment (nayanavenkataramana) [reply src/DeviceController.cpp]: replied above

File: src/DeviceController.cpp (hunk 4 above)
Comment (litcoder) [reply src/DeviceController.cpp]: Understand that this is for making sure other configurations. What I would like to suggest here is figuring out whether this is necessary. If you tested the single WL waiting code on SOS environment and it didn't work it leaves us no choice but if it did, I see no reason to take future maintenance risks.

**Comment by nayanavenkataramana** [comment]
> 1. As you're adding code from media-ctl library which is LGPL2.1 in static linking, you may need to update OSPDT and get a review for the license compatibilities.
> 2. Consider factory pattern for multiple camera configuration supporting for a code maintenance.

**Pull request merged**